python/build/lib/pcc/dataframe.py

```diff
@@ -1,149 +1,835 @@
+        with self.lock:
+            for tp in types:
+                self.__add_type(tp)
+
+    def remove_type(self, tp):
+        with self.lock:
+            tpname = tp.__realname__
+            if tpname in self.name2class:
+                gp = self.member_to_group[tpname]
+                if gp == tpname and len(self.group_to_members[gp]) > 1:
+                    raise TypeError("Cannot delete %s without removing all subtypes of it first" % tpname)
+                del self.member_to_group[tpname]
+                self.group_to_members[gp].remove(tp)
+                if tpname in self.object_map:
+                    del self.object_map[tpname]
+                del self.name2class[tpname]
+                del self.categories[tpname]
+                if tpname in self.impure:
+                    self.impure.remove(tpname)
             else:
-                for attr in dir(original_tp):
-                    if type(getattr(original_tp, attr)) == spacetime_property:
-                        container.__dimensions__.add(getattr(original_tp, attr))
-                        container.__dimensions_name__.add(attr)
-            for dimension in container.__dimensions__:        
-                value = getattr(obj, dimension._name)
-                setattr(newobj, dimension._name, self.__recurse_map[self.__get_type(value)](value))
-            newobj._start_tracking = True
-            return newobj
+                raise TypeError("Type %s is not in the dataframe." % tpname)
+
+    def append(self, tp, obj):
+        with self.lock:
+            groupname, id = self.__append(tp, obj)
+        self.__adjust_pcc({id, obj}, groupname)
+        self.apply_records_in_cache()
+        
+    def extend(self, tp, objs):
+        objmap = {}
+        with self.lock:
+            for obj in objs:
+                groupname, id = self.__append(tp, obj)
+                objmap[id] = obj
+        self.__adjust_pcc(objmap, groupname)
+        self.apply_records_in_cache()
+    
+    def get(self, tp, parameters=None):
+
+        tpname = tp.__realname__
+        if tpname not in self.observing_types:
+            raise TypeError("Type %s not registered" % tpname)
+        
+        with self.lock:
+            if not self.__is_impure(tp, self.categories[tp.__realname__]) and tpname in self.object_map:
+                return self.object_map[tpname].values()
+
+        if not self.calculate_pcc:
+            # Not calculating pcc if Client/Cache mode
+            return []
+        groupkey = self.member_to_group[tpname]
+        objs = []
+        with self.lock:
+            objs = self.__calculate_pcc([tp], self.object_map, parameters)
+        return_values = self.__record_pcc_changes(objs, {}.fromkeys(objs, []))[tp]
+        self.apply_records_in_cache()
+        return return_values
+
+    def delete(self, tp, obj):
+        with self.lock:
+            self.__check_validity(tp)
+            # all clear to delete
+            id = obj.__primarykey__
```

Comment:
Again, this seems like user-accessible function, assuming that they are giving you a PCC object is not good. Check whether the object is valid and throw an appropriate exception